clarifai/client/runner.py

```diff
@@ -84,72 +91,134 @@ def _run(self, request: service_pb2.PostModelOutputsRequest) -> service_pb2.Mult
 
   def _long_poll_loop(self):
-    """This method will long poll for work, and when it gets work, it will run the model on the inputs
+    """
+    This method will long poll for work, and when it gets work, it will run the model on the inputs
     and post the results back to the Clarifai API. It will then long poll again for more work.
     """
-    c = 0
-    # TODO: handle more errors within this loop so it never stops.
-    # TODO: perhaps have multiple processes running this loop to handle more work.
-    while True:
-      # Long poll waiting for work.
-      self.logger.info("Loop iteration: {}".format(c))
-      request = service_pb2.ListRunnerItemsRequest(
-          user_app_id=self.user_app_id, runner_id=self.runner_id)
-      work_response = self._grpc_request(self.STUB.ListRunnerItems, request)
-      if work_response.status.code == status_code_pb2.RUNNER_NEEDS_RETRY:
-        c += 1
-        continue  # immediate restart the long poll
-      if work_response.status.code != status_code_pb2.SUCCESS:
-        raise Exception("Error getting work: {}".format(work_response.status.description))
-      if len(work_response.items) == 0:
-        self.logger.info("No work to do. Waiting...")
-        continue
-
-      # We have work to do. Run the model on the inputs.
-      for item in work_response.items:
-        if not item.HasField('post_model_outputs_request'):
-          raise Exception("Unexpected work item type: {}".format(item))
-        self.logger.info(
-            f"Working on item: {item.id} with inputs {len(item.post_model_outputs_request.inputs)}"
-        )
-        result = self._run(item.post_model_outputs_request)
-
-        request = service_pb2.PostRunnerItemOutputsRequest(
-            user_app_id=self.user_app_id,
-            item_id=item.id,
-            runner_id=self.runner_id,
-            runner_item_outputs=[service_pb2.RunnerItemOutput(multi_output_response=result)])
-        result_response = self._grpc_request(self.STUB.PostRunnerItemOutputs, request)
-        if result_response.status.code != status_code_pb2.SUCCESS:
-          raise Exception(
-              json_format.MessageToJson(result_response, preserving_proto_field_name=True))
-          # raise Exception("Error posting result: {}".format(result_response.status.description))
+
+    def _single_thread_loop():
+      c = 0
+      while True:
+        # Long poll waiting for work.
+        try:
+          self.logger.info(f"Loop iteration: {c} for thread {threading.get_ident()}")
+          work_response = self._grpc_request(self.STUB.ListRunnerItems,
+                                             service_pb2.ListRunnerItemsRequest(
+                                                 user_app_id=self.user_app_id,
+                                                 runner_id=self.runner_id))
+          if work_response.status.code == status_code_pb2.RUNNER_NEEDS_RETRY:
+            c += 1
+            continue  # immediate restart the long poll
```

Comment:
i don't think so because the needs retry status code comes after a timeout on the backend side when there is no work items to be done for this runner and the timeout is pretty long. The runner has to retry right away in case a low latency request comes in it needs to be responsive to that.